src/core/annotations/qgsannotationitem.h

```diff
@@ -0,0 +1,155 @@
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSANNOTATIONITEM_H
+#define QGSANNOTATIONITEM_H
+
+#include "qgis_core.h"
+#include "qgis_sip.h"
+#include "qgscoordinatereferencesystem.h"
+#include "qgsrendercontext.h"
+#include "qgslinestring.h"
+#include "qgspolygon.h"
+
+class QgsFeedback;
+class QgsMarkerSymbol;
+class QgsLineSymbol;
+class QgsFillSymbol;
+
+/**
+ * \ingroup core
+ * Abstract base class for annotation items which are drawn with QgsAnnotationLayers.
+ *
+ * \since QGIS 3.16
+ */
+class CORE_EXPORT QgsAnnotationItem
+{
+
+#ifdef SIP_RUN
+    SIP_CONVERT_TO_SUBCLASS_CODE
+    if ( sipCpp->type() == "marker" )
+    {
+      sipType = sipType_QgsAnnotationMarkerItem;
+    }
+    else if ( sipCpp->type() == "linestring" )
+    {
+      sipType = sipType_QgsAnnotationLineStringItem;
+    }
+    else if ( sipCpp->type() == "polygon" )
+    {
+      sipType = sipType_QgsAnnotationPolygonItem;
+    }
+    else
+    {
+      sipType = 0;
+    }
+    SIP_END
+#endif
+
+  public:
+
+    /**
+     * Constructor for an annotation item, with the specified \a crs for storing
+     * its geographic location.
+     */
+    QgsAnnotationItem( const QgsCoordinateReferenceSystem &crs );
+
+#ifndef SIP_RUN
+    //! QgsAnnotationItem cannot be copied
+    QgsAnnotationItem( const QgsAnnotationItem &other ) = delete;
+    //! QgsAnnotationItem cannot be copied
+    QgsAnnotationItem &operator=( const QgsAnnotationItem &other ) = delete;
+#endif
+
+    virtual ~QgsAnnotationItem() = default;
+
+    /**
+     * Returns a clone of the item. Ownership is transferred to the caller.
+     */
+    virtual QgsAnnotationItem *clone() = 0 SIP_FACTORY;
+
+    /**
+     * Returns a unique (untranslated) string identifying the type of item.
+     */
+    virtual QString type() const = 0;
+
+    /**
+     * Returns the bounding box of the item's geographic location.
+     *
+     * The coordinate reference system for the bounding box is retrieved via crs().
+     */
+    virtual QgsRectangle boundingBox() const = 0;
+
+    /**
+     * Returns the CRS used for storing the location of the item.
+     *
+     * \see setCrs()
+     */
+    QgsCoordinateReferenceSystem crs() const { return mCrs; }
+
+    /**
+     * Sets the \a crs used for storing the location of the item.
+     *
+     * \see crs()
+     */
+    void setCrs( const QgsCoordinateReferenceSystem &crs );
+
+    /**
+     * Renders the item to the specified render \a context.
+     *
+     * The \a feedback argument can be used to detect render cancelations during expensive
+     * render operations.
+     */
+    virtual void render( QgsRenderContext &context, QgsFeedback *feedback ) = 0;
```

Comment:
Yes -- all the symbology and text rendering code use modifiable references to the context, because they are modifying the expression context, painter, and other properties of the context.